libs/s25main/ingameWindows/iwObservate.cpp

```diff
@@ -69,53 +73,19 @@ void iwObservate::Msg_ButtonClick(const unsigned ctrl_id)
                 view->SetZoomFactor(2.3f);
             break;
         case 2:
-        {
-            if(followMovableId)
-                followMovableId = 0;
-            else
-            {
-                const DrawPoint centerDrawPt = DrawPoint(view->GetSize() / 2u);
-
-                double minDistance = std::numeric_limits<double>::max();
-
-                for(int y = view->GetFirstPt().y; y <= view->GetLastPt().y; ++y)
-                {
-                    for(int x = view->GetFirstPt().x; x <= view->GetLastPt().x; ++x)
-                    {
-                        Position curOffset;
-                        const MapPoint curPt =
-                          view->GetViewer().GetTerrainRenderer().ConvertCoords(Position(x, y), &curOffset);
-                        DrawPoint curDrawPt = view->GetWorld().GetNodePos(curPt) - view->GetOffset() + curOffset;
-
-                        if(view->GetViewer().GetVisibility(curPt) != Visibility::Visible)
-                            continue;
-
-                        for(const noBase& obj : view->GetWorld().GetFigures(curPt))
-                        {
-                            const auto* movable = dynamic_cast<const noMovable*>(&obj);
-                            if(!movable)
-                                continue;
-
-                            DrawPoint objDrawPt = curDrawPt;
-
-                            if(movable->IsMoving())
-                                objDrawPt += movable->CalcWalkingRelative();
-
-                            DrawPoint diffToCenter = objDrawPt - centerDrawPt;
-                            double distance = sqrt(pow(diffToCenter.x, 2) + pow(diffToCenter.y, 2));
-
-                            if(distance < minDistance)
-                            {
-                                followMovableId = movable->GetObjId();
-                                minDistance = distance;
-                            }
-                        }
-                    }
-                }
-            }
-
+            if(following)
+                // Stop following
+                pickedObject.Invalidate();
+            else if(!pickedObject.IsValid())
+                // If object is invalid, pick new object at center of view
+                pickedObject = PickedMovableObject::PickAtViewCenter(*view, false);
+
+            // Follow picked object, if valid
+            following = lastValid = pickedObject.IsValid();
+            // Cancel expiration; only has an effect if we started following the initially picked object
+            pickedObject.CancelExpiration();
```

Comment:
DRY: 
```suggestion
            pickedObject.CancelExpiration(); // only has an effect if we started following the initially picked object
```

Or better: Cancel only if pickedObject was valid at the if before

Reply:
Forcing it into the if before seems less intuitive, and canceling is cheap.